pkg/controller/updatechannel/updatechannel.go

```diff
@@ -0,0 +1,184 @@
+}
+
+func (h *handler) ClearPolling(key string, res *v1.UpdateChannel) (*v1.UpdateChannel, error) {
+	logrus.Debugf("### K3OS::UPDATE-CHANNEL >> ClearPolling >> node=%s >> key=%q >> spec=%+v >> status=%+v", h.node, key, res.Spec, res.Status)
+	defer logrus.Debugf("### K3OS::UPDATE-CHANNEL << ClearPolling << node=%s << key=%q << spec=%+v << status=%+v", h.node, key, res.Spec, res.Status)
+
+	res.Status.Polling = ""
+	return res, nil
+}
+
+func (h *handler) SetUpgrading(key string, res *v1.UpdateChannel) (*v1.UpdateChannel, error) {
+	logrus.Debugf("### K3OS::UPDATE-CHANNEL >> UpdateUpgrading >> node=%s >> key=%q >> spec=%+v >> status=%+v", h.node, key, res.Spec, res.Status)
+	defer logrus.Debugf("### K3OS::UPDATE-CHANNEL << UpdateUpgrading << node=%s << key=%q << spec=%+v << status=%+v", h.node, key, res.Spec, res.Status)
+
+	res.Status.Upgrading = append(res.Status.Upgrading, h.node)
+	return res, nil
+}
+
+func (h *handler) ClearUpgrading(key string, res *v1.UpdateChannel) (*v1.UpdateChannel, error) {
+	logrus.Debugf("### K3OS::UPDATE-CHANNEL >> ClearUpgrading >> node=%s >> key=%q >> spec=%+v >> status=%+v", h.node, key, res.Spec, res.Status)
+	defer logrus.Debugf("### K3OS::UPDATE-CHANNEL << ClearUpgrading << node=%s << key=%q << spec=%+v << status=%+v", h.node, key, res.Spec, res.Status)
+
+	upgrading := make([]string, len(res.Status.Upgrading))
+	for _, node := range res.Status.Upgrading {
+		if node != h.node {
+			upgrading = append(upgrading, node)
+		}
+	}
+	res.Status.Upgrading = upgrading
+	return res, nil
+}
+
+func (h *handler) PollLatest(key string, res *v1.UpdateChannel) (*v1.UpdateChannel, error) {
+	logrus.Debugf("### K3OS::UPDATE-CHANNEL >> UpdateLatest >> node=%s >> key=%q >> spec=%+v >> status=%+v", h.node, key, res.Spec, res.Status)
+	defer logrus.Debugf("### K3OS::UPDATE-CHANNEL << UpdateLatest << node=%s << key=%q << spec=%+v << status=%+v", h.node, key, res.Spec, res.Status)
+
+	channel, err := upgrade.NewChannel(res.Spec.URL)
+	if err != nil {
+		return res, err
+	}
+	latest, err := channel.Latest()
+	if err != nil {
+		return res, err
+	}
+	res.Spec.Version = latest.Name
+	return res, nil
+}
+
+func (h *handler) UpgradeSystem(key string, res *v1.UpdateChannel) (*v1.UpdateChannel, error) {
+	logrus.Debugf("### K3OS::UPDATE-CHANNEL >> UpgradeSystem >> node=%s >> key=%q >> spec=%+v >> status=%+v", h.node, key, res.Spec, res.Status)
+	defer logrus.Debugf("### K3OS::UPDATE-CHANNEL << UpgradeSystem << node=%s << key=%q << spec=%+v << status=%+v", h.node, key, res.Spec, res.Status)
+
+	cmd := exec.Command("k3os-upgrade", "download-rootfs")
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		logrus.Warnf("%q: %s", cmd.Path, err.Error())
+		return res, err
+	}
+	logrus.Infof("%q: %s", cmd.Path, out)
+
+	cmd = exec.Command("k3os-upgrade", "download-kernel")
+	out, err = cmd.CombinedOutput()
+	if err != nil {
+		logrus.Warnf("%q: %s", cmd.Path, err.Error())
+		return res, err
+	}
+	logrus.Infof("%q: %s", cmd.Path, out)
```

Comment:
@ibuildthecloud these look to be taking too long such that the operator gets into a restart state. these should be proper jobs i think?

Reply:
refactored to use jobs